### ISSUE TITLE
fix video issues

### ISSUE THREAD

**Issue by m3t4f1v3** [m3t4f1v3/onscripter-ru#2]
does not break videos

- [x] sdl2
- [x] sdl2_gpu
- [x] sd2_image
- [x] sdl2_mixer
- [x] bzip
- [ ] ffmpeg
- [x] freetype
- [x] fribidi
- [x] harfbuzz
- [x] jpeg
- [x] libass
- [x] libcxx
- [x] libcxxabi
- [x] libepoxy
- [x] libogg
- [x] libpng
- [x] libunwind
- [x] libusb
- [x] libvorbis
- [x] lua
- [x] smpeg2
- [x] zlib

**Comment by m3t4f1v3** [comment]
this should also clean up any badly implemented packages

**Comment by m3t4f1v3** [comment]
and building on other platforms

**Comment by m3t4f1v3** [comment]
if i get really lazy i may just make a thing that lets you use precompiled libs

**Comment by m3t4f1v3** [comment]
sdl2 needs to be cmake'd

**Comment by m3t4f1v3** [comment]
the text issue is probably due to sdl_gpu

**Comment by m3t4f1v3** [comment]
it definitely is

**Comment by m3t4f1v3** [comment]
:trollface: not doing macos or ios because I have no access to any of them

**Comment by m3t4f1v3** [comment]
also probably not fixing the libepoxy issue either for mingw

**Comment by vit9696** [comment]
Could you build CI on top of GitHub Actions please?

> not doing macos or ios because I have no access to any of them
Also, for macOS and iOS you could go with the CI builds, and I can test both as I use macOS primarily.

> also probably not fixing the libepoxy issue either for mingw
What is the issue?

I noticed that you hardcoded chapter names in the engine code. Could you please provide them through the script rather than hardcoding in the game?

Also, it is getting to the point where a lot of stuff starts working. Would you open the PR, so that your changes can be merged upstream?

**Comment by m3t4f1v3** [comment]
1. I will try
2. I've most likely figured out the fix (edit nvm, its now throwing undefined reference to 'glGetIntegerV')
3. I will do so, but I doubt there really will be a good way to adapt it to other visual novels
4. I will probably do so once I've finished off the discord rpc integration (my original goal)

**Comment by m3t4f1v3** [comment]
@vit9696 do you know of a good position to place a callback run that needs to be constantly refreshed, I've looked at ConstantRefresh.cpp but i cant exactly figure out how to hook into it

**Comment by vit9696** [comment]
Rather tell me what do you need, since the "position" will vary depending on the desires.

**Comment by m3t4f1v3** [comment]
So discord requires you to place a callback function on every tick in order for it to send it back to the client,
placing it in a while loop in event.cpp causes a segmentation failure but only running the callbacks on every set presence command causes it to not show up at all.

I think using the constant refresh actions is a good way to overcome it but I do not understand how it works.

**Comment by vit9696** [comment]
Hmm, and what does/should happen in the callback?

**Comment by m3t4f1v3** [comment]
it sends the rich presence information to the discord client

**Comment by vit9696** [comment]
Feels like you can just put it to `advanceGameState`… But you can also subclass `ConstantRefreshAction` and add code to `advance` method. Then you can add it like:

```C++
Lock lock(&ons.registeredCRActions);
ons.registeredCRActions.emplace_back(YourAction::create())
```

Due to the nature of such action (constant callback) I may have to ask you to make discord support optional:
- Permit not compiling with Discord at all (via a configure option)
- Do not enable it without an explicit user agreement (via the ons.cfg option)

This seems important to me for privacy reasons.

**Comment by m3t4f1v3** [comment]
placing it in advanceGameState still seems to cause segmentation faults and the user must use an explicitly patched version of the script for it to even send rpc updates

**Comment by vit9696** [comment]
Well, hmm, you then need to define what "tick" is. If it needs to be updated like every second, then you could create e.g. a `game_fps`-based counter that will trigger discord update every second or something like that.

**Comment by m3t4f1v3** [comment]
Discord themselves did not define it

I believe I've finished making the discord integration optional

**Comment by m3t4f1v3** [comment]
there also seems to be something wrong with either the subtitles or loading videos
the error message is `***[Warning] Parse Issue at line 568333 (*pam:19) - failed to load the video ***
	(String buffer: [video/sub/both_eng.ass])`

**Comment by vit9696** [comment]
Does the file exist though?

**Comment by m3t4f1v3** [comment]
yes, I have rechecked it with the hashes and it returns without error

**Comment by m3t4f1v3** [comment]
It also does not work even if you disable subtitles in the configuration

**Comment by vit9696** [comment]
Then ffmpeg or libass (integration) is broken.

**Comment by m3t4f1v3** [comment]
I have tried reverting both to the original version but it still does not work

**Comment by vit9696** [comment]
Well, try the actual failure from loadVideo onwards. In my opinion, even if the old version does not work, it changes little. Can be build environment error.

**Comment by m3t4f1v3** [comment]
running under gdb, overriding the retcontinue, I get `Thread 1 "onscripter-ru" received signal SIGSEGV, Segmentation fault.
0x000055555574b570 in MediaProcController::frameSize(SDL_Rect const&, int&, float&, int&, float&, bool) ()`

**Comment by m3t4f1v3** [comment]
I've completely fixed the discord rpc issue, although it is a bit hardcoded
I still have no clue as to what causes the video issue however

**Comment by vit9696** [comment]
Well, put a bp on movieCommand and find why exactly loadVideo fails.

**Comment by m3t4f1v3** [comment]
looks like its at `if (!hasStream(VideoEntry))`

**Comment by vit9696** [comment]
Then it means `MediaProcController::findDecoder` fails. Which likely means your ffmpeg is compiled incorrectly.

**Comment by vit9696** [comment]
CI fails to build due to enabled tiff codec apparently. Try adding `--disable-tiff`.

**Comment by m3t4f1v3** [comment]
I reinstalled my entire os and it seems that the video issue isnt an environment error

**Comment by m3t4f1v3** [comment]
most likely due to a hacky fix for 5.0 that removed the AVLockOp lockManager entirely

I'll be changing it down to 4.4 for now

**Comment by m3t4f1v3** [comment]
I'll attempt to test it on macos by using a virtual machine
